### PR TITLE
Docs extension system

### PR DESCRIPTION
**Pull request by kulmann**
## Description
Introducing a first iteration on developer documentation for our extension system. Collaborative effort with @pascalwengerter 

## Related Issue
- Works towards https://github.com/owncloud/web/issues/9892
- Supersedes https://github.com/owncloud/web/pull/10067

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests
- [x] Documentation
- [ ] Maintenance (e.g. dependency updates or tooling)

### CODE REVIEW

File: docs/extension-system/_index.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      extensions,
+      navItems,
+      routes
+    }
+  }
+})
+```
+
+By defining an application via `defineWebApplication` you can provide the following:
+- `appInfo` - the application metadata, which is used to make the application available via the app switcher and the app registry
```
Comment (JammingBen) [inline docs/extension-system/_index.md:81]: Should we link to the `ApplicationInformation` interface here? Or is this supposed to be a quick overview only?

```suggestion
- `appInfo` - the application metadata, which is used to make the application available via the app switcher and the app registry.
```

File: docs/extension-system/_index.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+- `navItems` - the statically defined navigation items for the left sidebar. Only gets rendered when more than 1 navigation item exists at runtime. 
+Additional dynamic navigation items can be registered via the extension registry.
+- `routes` - the routes to the different views of your application. May be referenced within the `navItems`. Authentication requirements can be defined per item.
```
Comment (JammingBen) [inline docs/extension-system/_index.md:84]: Same question here?

File: docs/extension-system/_index.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+#### Extension Points
+
+There are standardized components and places where extensions are being used automatically. These are the ones that are currently provided:
+
+1. Left Sidebar for Navigation
+2. Right Sidebar in any file(s) context 
+3. Folder Views in the files app 
+4. Right click context menu in the files app 
+5. Batch actions in the files app
+6. Search results in the search app
```
Comment (JammingBen) [inline docs/extension-system/_index.md:146]: This list is missing:

* `Upload menu in the files app`
* `Quick actions in the files list`

**Review by JammingBen — COMMENTED** [review]
Awesome stuff! 🥳  Just a few annotations from my side.

Also, one aspect which we missed so far is extending content via portals. See e.g. https://github.com/cernbox/web-extensions/pull/25/files#diff-af11596391b8427b4db2f25ddc99628189bbadd1227494c6e1ef24543cb4bd76 where additional content is being added to existing sidebar panels. The available portal-targets currently are:

* `app.files.sidebar.sharing.shared-with.top`
* `app.files.sidebar.sharing.shared-with.top`
* `app.files.sidebar.file.details.table`

File: docs/extension-system/extension-types/search.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+The listSearch object consists of:
+
+- `component` - Vue component that can render the values from the SearchResult below
+- `search(term: string)` - Function that exectues the search, based on a given term. The term is formatted in [KQL](https://owncloud.dev/services/search/#query-language). Please note that the returned values needs to be formatted to fit either `SearchResource`  or `GenericSearchResultItem` type
+
+#### PreviewSearch
+
+The previewSearch object extends the listSearch with one additional attribute:
+
+- `available` - Indicates whether a preview underneath the search bar is available for this search provider
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply docs/extension-system/extension-types/search.md:57]: I feel we should get rid of that ... 😂

File: docs/extension-system/_index.md (hunk 1 above)
Comment (kulmann) [reply docs/extension-system/_index.md]: For now only supposed to be a quick overview. We want to come up with more details for applications, when we write about different examples for different apps (e.g. viewers/editors vs. "extension host apps" vs. custom business logic apps, etc).

File: docs/extension-system/_index.md (hunk 3 above)
Comment (kulmann) [reply docs/extension-system/_index.md:146]: added 👍

File: docs/extension-system/_index.md (hunk 2 above)
Comment (kulmann) [reply docs/extension-system/_index.md:84]: Same answer as above :-) Only meant to be a quick overview for now.

File: docs/extension-system/extension-types/search.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kulmann) [reply docs/extension-system/extension-types/search.md:57]: Buuuut.... we're using it for CERN

[the remark this answers is not in the document]
Comment (kulmann) [reply docs/extension-system/extension-types/search.md:57]: CERN has a search provider which only goes to the search result page and doesn't have a search preview.

**Review by JammingBen — APPROVED** [review]

**Pull request merged**